### PR TITLE
Add Context Structure to Affect State Dependent Liftings

### PR DESCRIPTION
**Pull request by 2over12**

### CODE REVIEW

File: include/remill/Arch/Arch.h [hunk 1]
```diff
@@ -281,14 +285,16 @@ class Arch {
   //            walk up, one byte at a time, to `MaxInstructionSize(false)`
   //            bytes being passed to the decoder, until you successfully decode
   //            or ultimately fail.
-  virtual bool DecodeInstruction(uint64_t address, std::string_view instr_bytes,
-                                 Instruction &inst) const = 0;
+  virtual std::optional<DecodingContext::ContextMap>
```
Comment (lkorenc) [inline include/remill/Arch/Arch.h:288]: Introduce type alias?

File: include/remill/Arch/Arch.h [hunk 2]
```diff
@@ -281,14 +285,16 @@ class Arch {
...
+  DecodeInstruction(uint64_t address, std::string_view instr_bytes,
+                    Instruction &inst, DecodingContext context) const = 0;
```
Comment (lkorenc) [inline include/remill/Arch/Arch.h:290]: `context` is taken by value - is that an intentional choice and we expect to always take care of ownership? Or do we expect to mutate it? (Just trying to understand if `const &` is not a better choice).

File: include/remill/Arch/Arch.h [hunk 3]
```diff
@@ -281,14 +285,16 @@ class Arch {
...
 
   // Decode an instruction that is within a delay slot.
-  bool DecodeDelayedInstruction(uint64_t address, std::string_view instr_bytes,
-                                Instruction &inst) const {
+  std::optional<DecodingContext::ContextMap>
+  DecodeDelayedInstruction(uint64_t address, std::string_view instr_bytes,
+                           Instruction &inst, DecodingContext context) const {
     inst.in_delay_slot = true;
-    return this->DecodeInstruction(address, instr_bytes, inst);
+    return this->DecodeInstruction(address, instr_bytes, inst, context);
```
Comment (lkorenc) [inline include/remill/Arch/Arch.h:297]: Copy on `context`, move it.

File: include/remill/Arch/ArchBase.h [hunk 4]
```diff
@@ -31,13 +31,27 @@ namespace remill {
 
 struct Register;
 
-// Internal base architecture for all Remill-internal architectures.
-class ArchBase : public remill::Arch {
- protected:
+
+/// A simple mixin for a default decoder with a null context and the default insdtruction lifter;
+class DefaultContextAndLifter : virtual public remill::Arch {
```
Comment (lkorenc) [inline include/remill/Arch/ArchBase.h:36]: `virtual` inheritance is a big hammer we may not need here - think how to make it without.

File: include/remill/Arch/Context.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  DecodingContext() = default;
+
+  DecodingContext(std::unordered_map<std::string, uint64_t> context_value);
+
+
+  uint64_t GetContextValue(const std::string &context_reg) const;
+  DecodingContext PutContextReg(std::string creg, uint64_t value) const;
+
+  static ContextMap UniformContextMapping(DecodingContext cst);
```
Comment (lkorenc) [inline include/remill/Arch/Context.h:46]: What is the advantage of using this compared to simply passing the "raw" `unordered_map` around?

File: lib/Arch/Context.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+DecodingContext DecodingContext::PutContextReg(std::string creg,
+                                               uint64_t value) const {
+  std::unordered_map<std::string, uint64_t> new_value(this->context_value);
+  new_value.emplace(creg, value);
+  return DecodingContext(std::move(new_value));
+}
+
+DecodingContext::ContextMap
+DecodingContext::UniformContextMapping(DecodingContext cst) {
+  return [cst](uint64_t ctx) -> DecodingContext { return cst; };
```
Comment (lkorenc) [inline lib/Arch/Context.cpp:32]: `ctx` is simply ignored?

File: lib/BC/TraceLifter.cpp [hunk 7]
```diff
@@ -345,7 +345,9 @@ bool TraceLifter::Impl::Lift(
 
       inst.Reset();
 
-      (void) arch->DecodeInstruction(inst_addr, inst_bytes, inst);
+      // TODO(Ian): not passing context around in trace lifter
+      (void) arch->DecodeInstruction(inst_addr, inst_bytes, inst,
```
Comment (lkorenc) [inline lib/BC/TraceLifter.cpp:349]: `std::ignore = value_dont_care_about()`

**Review by lkorenc — CHANGES_REQUESTED** [review]

File: lib/Arch/Context.cpp (hunk 6 above)
Comment (lkorenc) [reply lib/Arch/Context.cpp:32]: you are doing two copies, move `cst` into lambda.

File: lib/Arch/Arch.cpp [hunk 8]
```diff
@@ -850,16 +850,30 @@ const IntrinsicTable *ArchBase::GetInstrinsicTable(void) const {
...
+std::optional<DecodingContext::ContextMap>
+DefaultContextAndLifter::DecodeInstruction(uint64_t address,
+                                           std::string_view instr_bytes,
+                                           Instruction &inst,
+                                           DecodingContext context) const {
   inst.SetLifter(std::make_unique<remill::InstructionLifter>(
       this, this->GetInstrinsicTable()));
-  return this->ArchDecodeInstruction(address, instr_bytes, inst);
+  if (this->ArchDecodeInstruction(address, instr_bytes, inst)) {
+    return [](uint64_t x) -> DecodingContext { return DecodingContext(); };
```
Comment (lkorenc) [inline lib/Arch/Arch.cpp:866]: Drop `x` - unused arguments do not need names.

File: lib/Arch/Sleigh/Arch.cpp [hunk 9]
```diff
@@ -366,28 +366,35 @@ std::string CustomLoadImage::getArchType(void) const {
...
+                              Instruction &inst,
+                              DecodingContext context) const {
   inst.SetLifter(
       std::make_shared<SleighLifter>(this, *this->GetInstrinsicTable()));
   assert(inst.GetLifter() != nullptr);
 
-  return this->ArchDecodeInstruction(address, instr_bytes, inst);
+  if (const_cast<SleighArch *>(this)->DecodeInstructionImpl(
+          address, instr_bytes, inst)) {
+    return [this](uint64_t x) -> DecodingContext {
```
Comment (lkorenc) [inline lib/Arch/Sleigh/Arch.cpp:379]: Drop name of `x`.

File: include/remill/Arch/Context.h (hunk 5 above)
Comment (2over12) [reply include/remill/Arch/Context.h:49]: The context map? the lambda allows the decoders to encode arbitrary computations over successor addresses

File: lib/Arch/Arch.cpp [hunk 10]
```diff
@@ -863,7 +863,7 @@ DefaultContextAndLifter::DecodeInstruction(uint64_t address,
   inst.SetLifter(std::make_unique<remill::InstructionLifter>(
       this, this->GetInstrinsicTable()));
   if (this->ArchDecodeInstruction(address, instr_bytes, inst)) {
-    return [](uint64_t x) -> DecodingContext { return DecodingContext(); };
+    return [](uint64_t _) -> DecodingContext { return DecodingContext(); };
```
Comment (lkorenc) [inline lib/Arch/Arch.cpp:864]: Just `[](uint64_t) -> DecodingContext { ... }`.
No need for `_`. (In other changes as well)

As a sidenote, `-> DecodingContext` can also be dropped, I think compiler should be able to deduce this, but I can understand if you want to keep it as part of docs.

**Review by lkorenc — CHANGES_REQUESTED** [review]

File: include/remill/Arch/ArchBase.h [hunk 11]
```diff
@@ -122,4 +105,27 @@ class ArchBase : virtual public remill::Arch {
   mutable std::unique_ptr<IntrinsicTable> instrinsics{nullptr};
 };
 
+class DefaultContextAndLifter : virtual public remill::ArchBase {
```
[the remark this answers is not in the document]
Comment (2over12) [reply include/remill/Arch/ArchBase.h:108]: So this change pushed down the virtual inheritance by one but there is still a diamond problem here due to shared structure in the AArch32 and x86 arches. ie. we have AArch32Base -> ArchBase, then we have Aarch32 (the non sleigh version) -> DefaultContextAndLifter and Aarch32 -> ArchBase, hence the virtual inheritance.

[the remark this answers is not in the document]
Comment (2over12) [reply include/remill/Arch/ArchBase.h:108]: The core problem I guess with those setups is the base class needs access to ArchBase that the sleigh arches are then going to derive from directly while another child needs the DefaultContextAndLifter which has to have access to ArchBase as well

**Review by lkorenc — APPROVED** [review]

**Pull request merged**